### DISCUSSION TITLE
Auto-Compact LOH on OutOfMemory Exception, or other trigger

### DISCUSSION THREAD

**Discussion by najak3d** [dotnet/runtime#65023] [Ideas]
This question ties to this post:
[https://github.com/dotnet/runtime/discussions/47659](https://github.com/dotnet/runtime/discussions/47659)

which references this partial fix:
[https://docs.microsoft.com/en-us/dotnet/core/runtime-config/garbage-collector#heap-limit-percent](https://docs.microsoft.com/en-us/dotnet/core/runtime-config/garbage-collector#heap-limit-percent)

But leaves much to be desired, as the result is tedious/non-elegant code to ensure you never get an OOM (or ASP restart when memory exceeds a limit).

What I would suggest is that you add a GC setting like this:

  **AutoCompactLargeObjectHeapOnOutOfMemoryException = true**

**If 'true', then when an OutOfMemory exception occurs (OOM), it would do a GC2 and Compact the LOH, and then would RETRY the allocation that failed!**

How useful would this be?!?!?   We can have 1000's of places in our code that might make the "final allocation that causes the OOM" - and we don't want to have 1000 mechanisms for "try allocation, catch OOM, Compact LOH, then retry allocation".   It would be FAR more elegant if the .NET runtime would simply do this for us!...   then if after LOH Compaction, and OOM occurs again, then it would actually throw the exception, and maybe crash your app.

Otherwise, we're forced to "over compensate" and have many forced Compaction of LOH, "just-in-case".  And that is so non-elegant and not nearly as effective as simply having .NET runtime do this for us.

Essentially, this feature would automatically change many CRITICAL exceptions, into a simple "few second hiccup" at worst, while alleviating much stress from memory-intensive .NET apps.

**Is this something the .NET runtime could feasibly do for us?**

**Comment by danmoseley** [comment]
cc @mangod9

**Reply by najak3d** [reply]
Can you be more explicit?  I'm assuming you are insinuating that you've solved this problem in your own fork of the runtime?  More detail would be appreciated.

**Reply by danmoseley** [reply]
@najak3d no, I was tagging @mangod9 as his team owns the GC 😸

**Reply by msedi** [reply]
@najak3d: I agree, I'm having the same problem. It seems that you really need to enfore compaction otherwise you are running into OOM. I'm not quite sure if I'm correct but in most threads (also here) I have read that it shouldn't be necessary because the GC does this automatically in high load conditions.

In my case I now when I'm cleaning up a lot of memory (here, removing LOH objects from a list). At least for me it would be good to enclose this in some `BeginCleanupRegion/EndCleanupRegion` where the GC might explicitly collect cleanup information, something like its already available if you allocate a lot of unmanaged memory (`AddMemoryPressure/RemoveMemoryPressure`).

**Reply by mangod9** [reply]
Yeah compaction should be happening when possible during low memory conditions, unless there are lot of pinned objects which are prohibiting relocation of objects. Have you been able to determine that compaction is not being triggered for some reason? I would suggest taking a trace which can help diagnose what might be happening specifically for your application.

**Reply by najak3d** [reply]
@mangod9  - the core issue is this -- we don't know when an OOM is going to occur.   And if we don't have the "CompactLOHOnce" flag on, it won't compact the LOH.   And so once we encounter a few OOM's, the natural response we have is to "get paranoid" and ALWAYS set "CompactLOHOnce" repeatedly -- to reduce the chance of getting an OOM.   And so then our app might be plagued with unnecessary hiccups (as LOH compaction takes a while).

And so if we do NOT rampantly/constantly set the "CompactLOHOnce" Flag (for safety, just-in-case), then the GC's that happen under duress will not compact the LOH.   I don't think "rampantly/constantly" setting this flag is a "good solution".

It would be MUCH better to simply set a flag ONCE at startup that tells the runtime to "AutoCompactLargeObjectsOnMemoryException" - so that we can stop ALL paranoia and risk -- it would literally ONLY compact the LOH if an OOM occurred.  And thus, it replaces for us, automagically, a CRITCAL exception (likely to cause malfunction, maybe total crash) with a simple 0.1 to 2 second hiccup.   And it ONLY happens when things get ugly, on demand.  No more paranoia, or over-LOH-Compaction.

Then the only OOM that we'd get is for the case where there truly is no-more-memory left (after compacting LOH), which would be fully acceptable (unless you offered another flag named "UseDiskCacheOnMemoryException" which would log this critical issue, but would slow your program immensely if it trashes to/from this disk cache."

OOM's are particularly insidious, because you can't really predict when/where they are going to happen, and so they often lead to CRITICAL failures.   Being able to more easily avoid OOM's should be a central objective of .NET runtime, IMO.

**Reply by najak3d** [reply]
Using DiskCache would actually be a FINE solution, so long as it notifies us of the condition (give us an event to register for) -- and then we'll MANUALLY force LOH compaction, which should then pull that allocation from the LOH.   In this instance, the hiccup that results would maybe be an extra second, added by the write/read to/from disk cache for the one allocation.

So if automating the "catching OOM, compacting LOH, then retrying the allocation", then maybe using DiskCache as an alternative to the OOM -- and notify us, so that we can manually handle this -- that would also be 10000% better than the current paranoia that all memory intensive .NET apps face now, without a good alternative.

**Reply by msedi** [reply]
@najak3d: I've tried the diskcache with memory mapped files. The problem is that in my situation I have two concurrent processes: 1. my process the produces/creates memory. 2. the memory manager. I'm producing obviously too much memory in a very short time so that the MM is not able to follow and I finally get in the same condition as coming in low memory LOH conditions and even worse the MM fills the whole physical memory, then enters the page file and ultimately halts the whole system. 

Finally I had to the conclusion theat also MM-files are not sufficient and I had to include my own throttling to given the MM enough time to page the data. But even then I had no possibility to find out that I'm back in good memory conditions again. So I had to call WinAPI methods (FlushFileBuffers and EmptyWorkingSet) to get back to a good condition.

I currently left trying MM-Files because I had no good control over the paging system, too much WinAPI was involved and I wanted to be platform independent. So I currently created my own memory management for this purpose. It works quite well but still having my own "cleanup thread" running that continously checks memory conditions and performs a cleanup (removing the LOH objects from the list). After the cleanup I have to run the well known LOH compaction.

Regarding the OOM have you tried the `MemoryFailPoints`. I did them one but I seemed to a bit out and was also slowing down things because it fails awkward allocating an object to catch the exception if the memory could be allocated.

**Reply by najak3d** [reply]
Thanks for chiming in @msedi.   That helps Microsoft gurus to understand the severity of this issue in .NET, and the utter sweetness and power of providing a built-in solution.

I wasn't suggesting a system "reliant upon MM files" (disk cache), but rather to allow us to configure .NET to resort to a MM file whenever it hits an OOM... and then fires an event that we can listen for, so that we know it happened.  The CRITICAL issue is avoiding the OOM EXCEPTION which blows up your app and flow.  The whole point of .NET is to make things easy/sensible.   And this is one area that REALLY should be handled elegantly by .NET runtime.

I can understand how it might be too difficult for .NET to:
1. Catch an OOM
2. Compact the LOH right there on-the-spot (waiting for all other threads to allow it -- which may not happen if there is a deadlock situation associated with the thread that induced the OOM)
3. And then retry the allocation.

I imagine the only way to make that happen, might be to do a partial-LOH compaction (compact whatever seems safe, to make at least ONE GAP that fits the new object.  But that's a lot of new code, for a sloppy result.

So, maybe if instead .NET just resorted to a MM file, to fill-the-gap, and then let the App manually handle the event, so that we can eliminate usage of the Disk-cache very quickly.   This seems like a LOT easier solution for the .NET runtime, but would be IMMENSELY USEFUL, to know that you know no longer need to be concerned about OOM's destroying your App's flow.

**Reply by msedi** [reply]
@najak3d. I agree something needs to be done. My problem is that I'm not aware if everything is currently running as expected and there is indeed a LOH compaction in low memory conditions. As @mangod9 said there already is a LOH compaction. The question is if it is too late in some conditions and you therefore get an OOM. But this is pretty hard to find out.

**Reply by najak3d** [reply]
Over the years, and even now, it seems that most .NET memory intensive programs have inappropriate/difficult issues with OOMs, and their answer is always to augment .NET's LOH with custom code, or paranoid logic.  But if .NET simply handled the OOM in a different manner, most of this concern disappears entirely.  All a .NET app *should* be concerned about is whether or not they are truly "using up all the RAM".

IMO, the runtime GC should have the following API:
1. bool CompactLOHOnce -- already exists.  Kudos!
2. bool UseDiskCacheOnOutOfMemory -   if true, .NET would use DiskCache instead of throwing an OOM exception.
3. event FullCollection  -- fires for every full collection that occurs.
4. event CacheAllocation -- fires each time disk cache is used for an allocation

Event Arguments for the above:
1.  FullCollection event:  
         TimeSpan TimeConsumed    (so you know the cost of the full collection - this way, if your app hiccups, you'll know the cause)
         long TotalBytesAllocated (includes small objects and large objects)
         long TotalBytesAvaliable
         long TotalBytesAvailabeIncludingCache -- this would give you the absolute Max before an OOM exception actually occurs
     (Also - perhaps these values could also be found statically, via  GC.LastFullCollectionEvent  - and see the same data.

2. CacheAllocation event:
        long CacheBytesAllocated 
        long TotalCacheBytesAllocated
        long TotalCacheBytesRemaining


Then, finally, .NET apps will be able to easily manage Memory, without writing excessive/paranoid/tedious code to augment .NET's own memory management logic.   As a personal .NET lover-to-the-max, it's always seemed a shame that .NET falters needlessly when it comes to supporting memory-intensive apps.

I am the original author, and co-founder of the Visual3D.NET engine (dead now, killed by Unity3D):

https://vimeo.com/8491336

(that's me narrating; intro was my daughter)

Our demise was exacerbated by .NET's poor support for high-memory-use apps.
